File: src/modules/AlgoDatCourse/implemented/python/TaskManager.py
import heapq

class Task:
    def __init__(self, name: str, priority: int):
        self.name = name
        self.priority = priority

    def __repr__(self):
        return f"[{self.priority}] {self.name}"


class TaskManager:
    """
    Pythons heapq ist wie Javas PriorityQueue intern ein MinHeap.
    Negierte Priorität simuliert einen MaxHeap (höchste Priorität zuerst).
    """
    def __init__(self):
        self._heap: list[tuple[int, int, Task]] = []
        self._counter = 0  # Tie-breaker bei gleicher Priorität

    def add_task(self, name: str, priority: int) -> None:
        task = Task(name, priority)
        heapq.heappush(self._heap, (-priority, self._counter, task))
        self._counter += 1

    def get_next_task(self) -> Task:
        return heapq.heappop(self._heap)[2]

    def peek_task(self) -> Task:
        return self._heap[0][2]

    def is_empty(self) -> bool:
        return len(self._heap) == 0


if __name__ == "__main__":
    manager = TaskManager()
    manager.add_task("E-Mails beantworten", 2)
    manager.add_task("Server neu starten", 10)
    manager.add_task("Hauspost wegbringen", 7)
    manager.add_task("Essensmarken besorgen", 1)
    manager.add_task("Backup durchführen", 5)

    print("Nächste Aufgabe:", manager.peek_task())
    print()
    while not manager.is_empty():
        print("Bearbeite:", manager.get_next_task())

File: src/modules/AlgoDatCourse/implemented/csharp/TaskManager.cs
using System;
using System.Collections.Generic;

public class Task
{
    public string Name;
    public int Priority;

    public Task(string name, int priority) { Name = name; Priority = priority; }

    public override string ToString() => $"[{Priority}] {Name}";
}

public class TaskManager
{
    // .NET 6+: PriorityQueue<TElement, TPriority>
    // Standardmäßig ein MinHeap (niedrigste TPriority zuerst).
    // Negierte Priorität als TPriority → höchste Priorität zuerst.
    private readonly PriorityQueue<Task, int> _queue = new();

    public void AddTask(string name, int priority)
        => _queue.Enqueue(new Task(name, priority), -priority);

    public Task GetNextTask() => _queue.Dequeue();
    public Task PeekTask()    => _queue.Peek();
    public bool IsEmpty       => _queue.Count == 0;

    public static void Main()
    {
        var manager = new TaskManager();
        manager.AddTask("E-Mails beantworten", 2);
        manager.AddTask("Server neu starten", 10);
        manager.AddTask("Hauspost wegbringen", 7);
        manager.AddTask("Essensmarken besorgen", 1);
        manager.AddTask("Backup durchführen", 5);

        Console.WriteLine("Nächste Aufgabe: " + manager.PeekTask());
        Console.WriteLine();
        while (!manager.IsEmpty)
            Console.WriteLine("Bearbeite: " + manager.GetNextTask());
    }
}
